### PR TITLE
feat(rust): create secret trait

### PR DESCRIPTION
**Pull request by SanjoDeundiak**

### CODE REVIEW

File: implementations/rust/kex/src/x3dh.rs [hunk 1]
```diff
@@ -104,21 +101,21 @@ const ENROLLMENT_MSG_SIZE: usize = 32 + 32 + 32 + 64 + 16;
 /// The responder of X3DH creates a prekey bundle that can be used to establish a shared
 /// secret key with another party that can use
 pub struct X3dhResponder {
-    identity_key: Option<SecretKeyContext>,
-    signed_prekey: SecretKeyContext,
-    one_time_prekey: SecretKeyContext,
+    identity_key: Option<Arc<Box<dyn Secret>>>,
```
Comment (mikelodder7) [inline implementations/rust/kex/src/x3dh.rs:104]: Can you add a comment why identity_key is Arc<Box> and the other keys are not?

File: implementations/rust/vault/src/ffi/mod.rs [hunk 2]
```diff
@@ -1,15 +1,44 @@
...
+impl Default for DefaultVaultSecretFfiConverter {
+    fn default() -> Self {
+        DefaultVaultSecretFfiConverter {}
+    }
+}
+
+impl SecretFfiConverter for DefaultVaultSecretFfiConverter {
+    fn secret_into_ffi(&self, secret: &Box<dyn Secret>) -> Result<u64, VaultFailError> {
+        // FIXME
+        Ok(DefaultVaultSecret::downcast_secret(secret).unwrap().0 as u64)
```
Comment (mikelodder7) [inline implementations/rust/vault/src/ffi/mod.rs:29]: Replace this FIXME

File: implementations/rust/vault/src/ffi/mod.rs [hunk 3]
```diff
@@ -1,15 +1,44 @@
...
+
+impl Default for DefaultVaultSecretFfiConverter {
+    fn default() -> Self {
+        DefaultVaultSecretFfiConverter {}
+    }
+}
+
+impl SecretFfiConverter for DefaultVaultSecretFfiConverter {
+    fn secret_into_ffi(&self, secret: &Box<dyn Secret>) -> Result<u64, VaultFailError> {
+        // FIXME
```
Comment (SanjoDeundiak) [inline implementations/rust/vault/src/ffi/mod.rs:28]: Remove

File: implementations/rust/vault/src/lib.rs [hunk 4]
```diff
@@ -1,5 +1,4 @@
 #![deny(
-    missing_docs,
```
Comment (mikelodder7) [inline implementations/rust/vault/src/lib.rs:2]: Lets change `missing_docs` to 

#![warn(missing_docs)]

File: implementations/rust/vault/src/software/mod.rs [hunk 5]
```diff
@@ -15,6 +17,34 @@ use std::collections::BTreeMap;
...
+    ) -> Result<&DefaultVaultSecret, VaultFailError> {
+        context
+            .downcast_ref::<DefaultVaultSecret>()
+            .map_err(|_| VaultFailErrorKind::InvalidSecret.into())
+    }
+}
+
+impl Secret for DefaultVaultSecret {}
+
+unsafe impl IntoFfi for DefaultVaultSecret {
```
Comment (SanjoDeundiak) [inline implementations/rust/vault/src/software/mod.rs:36]: Remove

**Review by mikelodder7 — APPROVED** [review]

**Review by robinbudd — APPROVED** [review]

**Pull request merged**